src/base/tests/test_util.py

```diff
@@ -389,6 +389,57 @@ def test_adapt_domain_view(self):
         self.assertIn("courriel", view_search_2.arch)
 
 
+class TestReplaceReferences(UnitTestCase):
+    def setUp(self):
+        self.cr.execute(
+            """
+            CREATE TABLE dummy_model(
+                             id serial PRIMARY KEY,
+                             res_id int,
+                             res_model varchar,
+                             extra varchar,
+                             CONSTRAINT uniq_constr UNIQUE(res_id, res_model, extra)
+                         );
+
+            INSERT INTO dummy_model(res_model, res_id, extra)
+                 VALUES -- the target is there with same res_id
+                        ('res.users', 1, 'x'),
+                        ('res.partner', 1, 'x'),
+
+                        -- two with same target and the target is there
+                        ('res.users', 2, 'x'),
+                        ('res.users', 3, 'x'),
+                        ('res.partner', 2, 'x'),
+
+                        -- two with same target and the target is not there
+                        ('res.users', 4, 'x'),
+                        ('res.users', 5, 'x'),
+
+                        -- target is there different res_id
+                        ('res.users', 6, 'x'),
+                        ('res.partner', 4, 'x')
+            """
+        )
+
+    def _ir_dummy(self, cr, bound_only=True):
+        yield util.IndirectReference("dummy_model", "res_model", "res_id")
+
+    def test_replace_record_references_batch__full_unique(self):
+        mapping = {1: 1, 2: 2, 3: 2, 4: 3, 5: 3, 6: 4}
+        with mock.patch("odoo.addons.base.maintenance.migrations.util.records.indirect_references", self._ir_dummy):
+            util.replace_record_references_batch(self.cr, mapping, "res.users", "res.partner")
```

Comment:
Because of the dynamic nature of the `odoo.addons.base.maintenance.migrations` package, we cannot mock it.

So we need to mock the `odoo.upgrade` version and limit the test to versions >= 13